### PR TITLE
Support Cumulocity Device Availability

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes

Todo:
- [x] Update all affected unit tests by this change (3764aa78b9df2c52727b43f3d8fa731b9b52215d)
- [x] Clarify two design decisions all regarding
- [x] Try the usecase: child metadata has the `leadService` field
- [x] Add dedicated unit tests for this feature
- [x] Add dedicated robot tests for this feature
- [x] Update the code according to the `@health` design
- [ ] Doc, probably in different PR as this PR is already huge

**"Lead Service" design decisions**
The "lead service" is meant to which service status `tedge-mapper-c8y` looks up to judge if the device is "available" or not.
- By default, using `te/device/{device_name}/service/tedge-agent/status/health`
- Can be customerized by device registration message with `@health` key. For example, 
Topic: `te/device/{device_name}//`
```json
{
  "@health": "can/be/any/custom/topic"
}
```
- If the health status is not reported to the given topic`@health`, `tedge-mapper-c8y` logs it as `info` and not sending a heartbeat.
- If the given topic is invalid as MQTT topic, `tedge-mapper-c8y` logs it as `info` and not sending a heartbeat.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2842

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 1]
```diff
@@ -979,10 +998,16 @@ impl CumulocityConverter {
         channel: Channel,
         message: &MqttMessage,
     ) -> Result<Vec<MqttMessage>, ConversionError> {
-        let mut registration_messages: Vec<MqttMessage> = vec![];
+        let mut converted_messages: Vec<MqttMessage> = vec![];
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1001]: All changes under this function `try_convert_te_topics` will be changed after merging https://github.com/thin-edge/thin-edge.io/pull/2884. Ignore the changes in this function for now.

File: tests/RobotFramework/tests/cumulocity/availability/heartbeat.robot [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+*** Test Cases ***
+
+### Main Device ###
+Heartbeat is sent
+    # How to test if a heartbeat is sent?
+    # c8y has time lag to change the availability state
+    # that may cause false-negative result if we check c8y_Availability.status
+    # also putting 'sleep 2m' a couple of times is not good :/
+
```
Comment (gligorisaev) [inline tests/RobotFramework/tests/cumulocity/availability/heartbeat.robot:18]: What about replacing Sleep commands with loops that periodically check the desired status. This approach can reduce wait times and improve test reliability. for Example something like:
`Wait Until Keyword Succeeds    2m    10s    Device Should Have Fragment Values    c8y_Availability.status=AVAILABLE`

File: crates/extensions/c8y_mapper_ext/src/actor.rs [hunk 3]
```diff
@@ -302,6 +321,66 @@ impl C8yMapperActor {
...
+                            &inventory_update_topic,
+                            &heartbeat_payload,
+                        ) {
+                            self.mqtt_publisher.send(message).await?;
+                        }
+                    }
+
+                    start_heartbeat_timer(
+                        self.timer_sender.clone(),
+                        interval,
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/actor.rs:370]: Using the same interval for the timer that we've set as the `c8y_RequiredAvailability` interval would lead to a lot of missed heartbeats. For e.g, if the interval is set to 30 mins and if we sleep for 30 minutes here using the timer, by the time we process the heartbeat response from the timer and send a response to the cloud, it definitely would take more than 30 mins, by then the availability interval in the cloud would have timed out. So, IMO, we should send heartbeat intervals more often than what the configured values is.

File: crates/extensions/c8y_mapper_ext/src/availability.rs [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::ServiceTopicId;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+use tracing::debug;
+use tracing::info;
+
+pub type TopicWithoutPrefix = String;
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/availability.rs:20]: Why not just use the `EntityTopicId` of the lead service, as the health channel info in them are just redundant?

File: tests/RobotFramework/tests/cumulocity/availability/c8y_required_availability.robot [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+Library    ThinEdgeIO
+
+Test Tags    theme:c8y
+Test Setup    Test Setup
+Test Teardown    Get Logs
+
+*** Test Cases ***
+
+c8y_RequiredAvailability is set by default
+    Execute Command    ./bootstrap.sh
```
Comment (albinsuresh) [inline tests/RobotFramework/tests/cumulocity/availability/c8y_required_availability.robot:13]:
```suggestion
    Setup
```
Why not just that I wonder.

File: crates/extensions/c8y_mapper_ext/src/availability.rs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        ))
+        .await?;
+
+    Ok(())
+}
+
+pub fn get_lead_service_topic(entity: &EntityMetadata) -> Result<TopicWithoutPrefix, String> {
+    match entity.other.get("@health") {
+        Some(maybe_topic_name) => match maybe_topic_name.as_str() {
+            Some(topic_name) if Topic::new(topic_name).is_ok() => Ok(topic_name.to_string()),
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/availability.rs:128]: I feel we should validate that the value is a valid entity topic id. I'd have validated that against the entity store as well, but may be that's too early, as the lead service may not be registered by then.

**Review by albinsuresh — CHANGES_REQUESTED** [review]

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 7]
```diff
@@ -239,6 +248,11 @@ pub struct CumulocityConverter {
     pub command_id: IdGenerator,
     // Keep active command IDs to avoid creation of multiple commands for an operation
     pub active_commands: HashSet<CmdId>,
+
+    // Keep the service statuses with topic name as a key for c8y_Availability monitoring
+    pub health_status: HashMap<TopicWithoutPrefix, HealthStatus>,
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:253]: It's different from `EntityTopicId`. `EntityTopicId` is restrict to 4 segments name like `device/main/a/b`, however, the requirement here is addressing custom topic without prefix. 

For example,
Original topic: `te/custom/anything`, `te/very/long/a/b/c/d/e/f`
- Invalid as EntityTopicId
- Valid as the topic to monitor health (`custom/anything`, `very/long/a/b/c/d/e/f`)

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 8]
```diff
@@ -967,6 +988,9 @@ impl CumulocityConverter {
     ) -> Result<Vec<MqttMessage>, ConversionError> {
         debug!("Mapping message on topic: {}", message.topic.name);
         trace!("Message content: {:?}", message.payload_str());
+
+        record_health_status(&self.mqtt_schema, &mut self.health_status, message);
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: Because `process_health_status_message` can address only `Channel::Health` topics, meaning, the topic must have `te/{valid_entity_topic_id}/status/health` scheme. But the requirement is to accept any topic as health endpoint as long as it has `te`(can be custom root) prefix.

File: crates/extensions/c8y_mapper_ext/src/availability.rs [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            debug!("Heartbeat message is not generated because the status of '{health_topic_name}' is not 'up'");
+            None
+        }
+    }
+}
+
+pub async fn set_heartbeat_timer_from_metadata(
+    period: i16,
+    status_map: &mut HashMap<TopicWithoutPrefix, HealthStatus>,
+    timer_map: &mut HashMap<EntityTopicId, TopicWithoutPrefix>,
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/availability.rs:53]: Exactly the intention is not to expand `converter.rs`. Fed up with scrolling the long lines and addressing conflicts when someone wants to refactor the file.

> If we had something like an AvailabilityManager struct with ownership of those maps in this module, operating on them independently, it would have been better.

We can discuss it later.

File: crates/extensions/c8y_mapper_ext/src/actor.rs [hunk 10]
```diff
@@ -45,13 +50,20 @@ use tedge_uploader_ext::UploadRequest;
...
+use tracing::debug;
 use tracing::error;
 use tracing::warn;
 
 const SYNC_WINDOW: Duration = Duration::from_secs(3);
 
-pub type SyncStart = SetTimeout<()>;
-pub type SyncComplete = Timeout<()>;
+pub type SyncStart = SetTimeout<TimeoutKind>;
+pub type SyncComplete = Timeout<TimeoutKind>;
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/actor.rs:60]: Ok, I'll rename it

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 11]
```diff
@@ -230,6 +238,7 @@ pub struct CumulocityConverter {
     pub auth_proxy: ProxyUrlGenerator,
     pub uploader_sender: LoggingSender<IdUploadRequest>,
     pub downloader_sender: LoggingSender<IdDownloadRequest>,
+    pub timer_sender: LoggingSender<SyncStart>,
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:241]: I think renaming `SyncStart` as you suggested [here](https://github.com/thin-edge/thin-edge.io/pull/2866#discussion_r1611199113) solves your point.

File: crates/extensions/c8y_mapper_ext/src/availability.rs (hunk 6 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/availability.rs:128]: The lead service can be anything, that's why the type is not `EntityTopicId` and also no querying to entity store.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: To support any random topic, that doesn't even have to be a thin-edge topic (not even with custom prefix and custom topic scheme), then we need to make the mapper subscribe to those random topics as well, so that it is aware of the health updates published to those topics. This would mean that we need to dynamically subscribe to those "open topics" as and when they are registered, which won't be easy as our MQTT actor currently doesn't support dynamic subscriptions. The other alternative would be for the mapper to subscribe to '#' and then process only the ones that are declared by the `@health` endpoints. Since the mapper has a persistent connection with the broker, it will unnecessarily increase the load on the broker to queue a lot of unnecessary data for the mapper, most of which won't even be processed.

So, my advice would be to restrict this feature to thin-edge health endpoints only for now (customers are still free to use custom prefixes and topic schemes), and expand it later to any random topics, once we add support for dynamic subscriptions. @reubenmiller Are you okay with that?

[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: No the `@health` value will still use the root_prefix (default being `te/`), so at least we need to subscribe for anything more than we are already doing.

The though process was to include the explict `/status/health` portion of the topic is more explicit, rather than having unexpected "status/health" added to the `@health` value.

File: tests/RobotFramework/tests/cumulocity/availability/c8y_required_availability.robot (hunk 5 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/availability/c8y_required_availability.robot:13]: `Setup` does a lot more than just bootstrapping a device. When using the docker adapter it will spawn a new container. The `Test Setup` is already creating the container (but with bootstrapping disabled), as the tests look to require fine-grain control over the initial connection.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: > No the @health value will still use the root_prefix (default being te/), so at least we need to subscribe for anything more than we are already doing.

So, it is guaranteed that the `@health` value would be the health endpoint of some entity, right? It could be using a custom topic scheme, and even a custom prefix, as long as tedge is also configured to use the same prefix. In that case, things are better. The PR description stated that the value `@health` value `can/be/any/custom/topic`, that doesn't even conform to the above mentioned requirements. My concerns were about trying to be so open to any random topic. As long as it is a valid thin-edge health status topic, we're good.

> The though process was to include the explict /status/health portion of the topic is more explicit, rather than having unexpected "status/health" added to the @health value.

I think there's some misunderstanding here. @rina23q implemented this contract with the assumption that we can't assume the `@health` value to be on the `health/status` channel, but rather can be any other random topic as well. If we're restricting the value to the the `health/status` channel, then all good. That's exactly what I was expecting as well.

> Though I guess if someone chooses to uses something like hello/world/status (which will be interpreted as te/hello/world/status, then thin-edge.io won't receive any messages from this topic (as it does not match any of the existing local mqtt topic subscriptions?

Correct. Since the mapper currently subscribes specifically to `te/+/+/+/+/<channel>/<data-type>` topics, Something published to a 4-part topic like `te/hello/world/status` won't even be received, as thin-edge expects at least a 5-part topic.


**TO SUMMARISE** my current understanding is as follows: The requirement was not to support any random topic with any prefix and any number of topic levels (e.g: `a/b/c/d/c/e/f/g/h` or `a/b/c`), but only to expect valid health status messages on explicit health status channel itself, using any valid custom prefix and topic scheme (the same that is configured with tedge). Correct me if this understanding is wrong.

[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: Yes the original requirement was to support any random topic (under the root_prefix). But since this opens up the problem that the mapper is not subscribing to a wildcard topic `{root_prefix}/#`, then you're right we won't get the notifications on the topic specified by `@health`.

Then let's change the requirement so that the `@health` should refer to an target entity (e.g. `device/main/sevice/tedge-agent` etc.)...then the agent implies that the `/status/health` topic suffix.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs [hunk 12]
```diff
@@ -493,7 +493,17 @@ define_tedge_config! {
...
+        },
+
+        availability: {
+            /// Enable sending heartbeat to Cumulocity. If false, it doesn't set c8y_RequiredAvailability and disable the heartbeat
+            #[tedge_config(example = "true", default(value = true))]
+            enable: bool,
+
+            /// Required interval, and period of the heartbeat only if enable is true
+            #[tedge_config(example = "60", default(value = 60_i16))]
+            period: i16,
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:505]: Also better to say this interval is expected to be a number of seconds and to use the tedge-config type [`Seconds`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_config/src/tedge_config_cli/models/seconds.rs#L11).

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
We need to avoid adding responsibilities to this already overly bloated `CumulocityConverter`.

Device availability should be a monitored by an actor that is fully independent of `C8yMapperActor` and `CumulocityConverter`. I would implement this feature in a fully independent extension.

The device availability monitor needs to subscribe for entity registration messages and heath topic for this entities (as defined by the `@health" property of their registration).
This actor doesn't need to share data with the c8y converter and have simply to store mapping as those proposed here (`timer_map`, `status_map`) to publish heartbeat at a regular pace.

The only part which is c8y related are the topics and format for these heartbeats.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:992]: > Then let's change the requirement so that the `@health` should refer to an target entity (e.g. `device/main/service/tedge-agent` etc.)...then the agent implies that the `/status/health` topic suffix. This means the `@health` property would follow be of a similar format to the `@parent` field that is used in custom entity registration messages.

I agree that this is the way to go. And not because the former desire to support random topics was adding too many technical constraints on the design.

Having the `@heath` property be an entity identifier makes sense and even more sense than a random topic:
- even if a child device is not running `tedge-agent`, this child device will have in any cases to define and populate `te` topics for its data and services: hence this is not restrictive
- the  `@heath` property is the identifier of the entity, more likely a service but not necessarily,  used to monitor the health of the device: this makes sense from a user perspective
- this is aligned to all the other proposals related to MQTT topics: easier to understand and implement.

**Comment by rina23q** [comment]
Move to draft status as I'm planning to re-write code from scratch. However, it's worth keeping unclosed until the whole work is done in order to keep the conversation history.

**Comment by didier-wenzek** [comment]
Closing as addressed by https://github.com/thin-edge/thin-edge.io/pull/2940
